test/add_or_update_test.dart

```diff
@@ -0,0 +1,143 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+// ignore_for_file: avoid_relative_lib_imports
+
+import 'package:test/test.dart' hide test, throws;
+import '../lib/realm.dart';
+
+import 'test.dart';
+
+part 'add_or_update_test.g.dart';
+
+@RealmModel()
+class _Friend {
+  @PrimaryKey()
+  late int id;
+  _Friend? bestFriend;
+  final friends = <_Friend>[];
+}
+
+Future<void> main([List<String>? args]) async {
+  await setupTests(args);
+
+  test('Transitive adds', () {
+    final r = getRealm(Configuration.local([Friend.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2, bestFriend: alice);
+
+    r.write(() => r.add(bob));
+    expect(bob.isManaged, true);
+    expect(alice.isManaged, true);
+    expect(r.find<Friend>(bob.id), bob);
+    expect(r.find<Friend>(alice.id), alice);
+  });
+
+  test('Transitive updates', () {
+    final r = getRealm(Configuration.local([Friend.schema]));
+
+    var alice = Friend(1);
+    var bob = Friend(2, bestFriend: alice);
+
+    r.write(() => r.add(bob));
+
+    final aliceAgain = Friend(1);
+    final bobAgain = Friend(2, bestFriend: aliceAgain);
```

Comment:
This test would fail, if not for the changes I made in this PR. It requires the update flag to be forwarded, when adding the linked object.

Reply:
My question is shouldn't we be updating the bestFriend field to something entirely new so it is clear the update is working. Now this one is "updating" to an object with the same primary key as before, thus it may not catch bugs if they happen.

Reply:
I can add another test as well